File: python-backend/app/api/endpoints/inventory.py
from typing import Any, List

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_active_user, get_db
from app.db.models import InventoryItem, Product, User
from app.schemas.inventory import InventoryAdjustment, InventoryItem as InventoryItemSchema

router = APIRouter(redirect_slashes=False)


@router.get("/", response_model=List[InventoryItemSchema])
def read_inventory(
    db: Session = Depends(get_db),
    skip: int = 0,
    limit: int = 100,
    current_user: User = Depends(get_current_active_user),
) -> Any:
    """
    Retrieve inventory items.
    """
    # Modificado para filtrar itens com product_id nulo
    inventory = db.query(InventoryItem).filter(InventoryItem.product_id.isnot(None)).offset(skip).limit(limit).all()
    
    # Verificação adicional para garantir que não haja valores nulos
    valid_inventory = []
    for item in inventory:
        if item.product_id is not None:
            valid_inventory.append(item)
    
    return valid_inventory


@router.post("/add", response_model=InventoryItemSchema)
def add_to_inventory(
    *,
    db: Session = Depends(get_db),
    adjustment: InventoryAdjustment,
    current_user: User = Depends(get_current_active_user),
) -> Any:
    """
    Add items to inventory.
    """
    # Ensure product exists
    product = db.query(Product).filter(Product.id == adjustment.product_id).first()
    if not product:
        raise HTTPException(status_code=404, detail="Product not found")
    
    # Check if inventory item already exists
    inventory_item = db.query(InventoryItem).filter(
        InventoryItem.product_id == adjustment.product_id
    ).first()
    
    if inventory_item:
        # Update existing inventory
        inventory_item.quantity += adjustment.quantity
    else:
        # Create new inventory item
        inventory_item = InventoryItem(
            product_id=adjustment.product_id,
            quantity=adjustment.quantity
        )
        db.add(inventory_item)
    
    db.commit()
    db.refresh(inventory_item)
    return inventory_item


@router.post("/remove", response_model=InventoryItemSchema)
def remove_from_inventory(
    *,
    db: Session = Depends(get_db),
    adjustment: InventoryAdjustment,
    current_user: User = Depends(get_current_active_user),
) -> Any:
    """
    Remove items from inventory.
    """
    # Ensure inventory item exists
    inventory_item = db.query(InventoryItem).filter(
        InventoryItem.product_id == adjustment.product_id
    ).first()
    
    if not inventory_item:
        raise HTTPException(status_code=404, detail="Product not in inventory")
    
    # Check if we have enough items
    if inventory_item.quantity < adjustment.quantity:
        raise HTTPException(status_code=400, detail="Not enough items in inventory")
    
    # Update inventory
    inventory_item.quantity -= adjustment.quantity
    
    db.commit()
    db.refresh(inventory_item)
    return inventory_item

File: node-backend/src/routes/inventory.js
const express = require('express');
const { InventoryItem, Product } = require('../models');
const { authenticateToken } = require('../middleware/auth');

const router = express.Router();

// Get inventory
router.get('/', authenticateToken, async (req, res) => {
  try {
    const { skip = 0, limit = 100 } = req.query;
    const inventory = await InventoryItem.findAll({
      offset: parseInt(skip),
      limit: parseInt(limit),
      include: [{
        model: Product,
        as: 'product'
      }]
    });
    res.json(inventory);
  } catch (error) {
    console.error('Get inventory error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Add to inventory
router.post('/add', authenticateToken, async (req, res) => {
  try {
    const { product_id, quantity } = req.body;

    const product = await Product.findByPk(product_id);
    if (!product) {
      return res.status(404).json({ error: 'Product not found' });
    }

    let inventoryItem = await InventoryItem.findOne({ where: { product_id } });
    
    if (inventoryItem) {
      inventoryItem.quantity += quantity;
      inventoryItem.last_updated = new Date();
      await inventoryItem.save();
    } else {
      inventoryItem = await InventoryItem.create({
        product_id,
        quantity,
        last_updated: new Date()
      });
    }

    res.json(inventoryItem);
  } catch (error) {
    console.error('Add inventory error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Remove from inventory
router.post('/remove', authenticateToken, async (req, res) => {
  try {
    const { product_id, quantity } = req.body;

    const inventoryItem = await InventoryItem.findOne({ where: { product_id } });
    if (!inventoryItem) {
      return res.status(404).json({ error: 'Inventory item not found' });
    }

    if (inventoryItem.quantity < quantity) {
      return res.status(400).json({ error: 'Insufficient inventory' });
    }

    inventoryItem.quantity -= quantity;
    inventoryItem.last_updated = new Date();
    await inventoryItem.save();

    res.json(inventoryItem);
  } catch (error) {
    console.error('Remove inventory error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;
